internal/resolver/oci.go

```diff
@@ -0,0 +1,153 @@
+/*
+Copyright The ORAS Authors.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package resolver
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"sync"
+
+	specs "github.com/opencontainers/image-spec/specs-go"
+	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
+	"oras.land/oras-go/v2/errdef"
+)
+
+const (
+	// OCIImageIndexFile is the file name of the index from the OCI Image Layout Specification.
+	// Reference: https://github.com/opencontainers/image-spec/blob/master/image-layout.md#indexjson-file
+	OCIImageIndexFile = "index.json"
+)
+
+// OCI is a resolver based on file system with the OCI-Image layout.
+// Reference: https://github.com/opencontainers/image-spec/blob/master/image-layout.md
+type OCI struct {
+	sync.Mutex
+	index  *ocispec.Index
+	refMap map[string]ocispec.Descriptor
+
+	indexPath string
+}
+
+// NewOCI creates a new OCI resolver.
+func NewOCI(root string) (*OCI, error) {
+	oci := &OCI{
+		indexPath: filepath.Join(root, OCIImageIndexFile),
+	}
+
+	err := oci.loadIndex()
+	if err != nil {
+		return nil, err
+	}
+
+	if err := oci.saveIndex(); err != nil {
+		return nil, err
+	}

```

Comment:
The question behind it is that `loadIndex` is a heavy operation that it need to read the `index.json` file and build the reference. It might not be efficient if we `loadIndex` each time we do read operations.